### DISCUSSION TITLE
What is 'ChildTask'?

### DISCUSSION THREAD

**Discussion by NickDarvey** [fsprojects/FSharp.Control.TaskSeq#189] [Q&A] [answered]
The README mentions parallelism is possible with `ChildTask`. What is `ChildTask`?

https://github.com/fsprojects/FSharp.Control.TaskSeq/blob/32c7a470846eca4852d323cc889a11061d41e7d5/README.md?plain=1#L199

Is the planned support (also mentioned) something like operators that let you specify the degree of paralellism?

**Comment by abelbraaksma** [comment, marked as answer]
@NickDarvey sorry for responding this late, I missed the notification on this thread somehow. 

The issue with making `IAsyncEnumerable<'T>` parallel is non-trivial. The reason is that the interface is deliberately designed in such a way that you must `await` the `MoveNext` action (which returns `true|false` signaling whether there are more elements) before you can get to the `Current` property.

In other words, calling `MoveNext` multiple times can lead parallel execution, sure, but it can also lead to reading _past the end of the stream_ and _returning items in a random order_. What's worse, is that, just like a normal `seq`, the moments that side effects are executed are very well defined. However, once you run this in parallel, this guarantee goes out of the window.

> What is ChildTask?

This term comes from `Async.StartChild` and `Async.StartChildAsTask`, which are part of the F# Core library. They start a child task and enable parallel execution. A good definition and a discussion of the concept can be found here: https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/attached-and-detached-child-tasks

------------

In general, I have been wanting to add (easy) parallelism to the library, but as you may have guessed, parallelism is never easy. `AsyncSeq` has some support for it, as does `Async.Parallel`, obviously. 

My current idea is to either take `AsyncSeq`, but then the parallel parts, and simply live with the fact that these concepts should be kept separate, or I find some way of extending the current implementation, and allow `taskSeq` to have parallel tasks.

On the latter, the only way I see this working is if there's a worker that tells us that we can `MoveNext` (i.e. that there are more items), but that it isn't finished with its task yet. 

It is easy to confuse an async sequence with a sequence of asyncs. The first is supposed to be executed in order, while the other can be executed in parallel, or at will, in any order.

PS: is there a specific use case you had in mind? It may help prioritize this.

**Reply by NickDarvey** [reply]
> The issue with making IAsyncEnumerable<'T> parallel is non-trivial. The reason is that the interface is deliberately designed in such a way that you must await the MoveNext action (which returns true|false signaling whether there are more elements) before you can get to the Current property.

That makes sense.

> PS: is there a specific use case you had in mind? It may help prioritize this.

I wish I'd documented the use case when I asked this question—because now I can't remember.
